File: waypoints-legacy-importer/src/main/java/de/md5lukas/waypoints/legacy/internal/LocationTag.java
package de.md5lukas.waypoints.legacy.internal;

import de.md5lukas.nbt.Tag;
import org.bukkit.Bukkit;
import org.bukkit.Location;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

class LocationTag extends Tag {

    private Location location;

    public LocationTag(String name, Location location) {
        super(name);
        this.location = location;
    }

    public LocationTag(String name) {
        super(name);
    }

    public Location value() {
        return location;
    }

    @Override
    public void write(DataOutput dos) throws IOException {
        dos.writeUTF(location.getWorld().getName());
        dos.writeDouble(location.getX());
        dos.writeDouble(location.getY());
        dos.writeDouble(location.getZ());
    }

    @Override
    public void load(DataInput dis) throws IOException {
        location = new Location(Bukkit.getWorld(dis.readUTF()), dis.readDouble(), dis.readDouble(), dis.readDouble());
    }

    @Override
    public String toString() {
        return location == null ? "null" : String.format("World: %s X: %f Y: %f Z: %f", location.getWorld() == null ? "null" : location.getWorld().getName(), location.getX(),
                location.getY(), location.getZ());
    }

    @Override
    public byte getId() {
        return 100;
    }

    @Override
    public String getTagName() {
        return "TAG_Location";
    }

    @Override
    public Tag copy() {
        return new LocationTag(getName(), location.clone());
    }
}

File: waypoints-legacy-importer/src/main/kotlin/de/md5lukas/waypoints/legacy/LocationTag.kt
package de.md5lukas.waypoints.legacy

import de.md5lukas.nbt.Tag
import org.bukkit.Bukkit
import org.bukkit.Location
import java.io.DataInput
import java.io.DataOutput

internal class LocationTag(
    name: String,
    private var location: Location? = null
) : Tag(name) {

    fun value(): Location {
        return location!!
    }

    override fun write(dos: DataOutput) {
        location!!.let {
            dos.writeUTF(it.world!!.name)
            dos.writeDouble(it.x)
            dos.writeDouble(it.y)
            dos.writeDouble(it.z)
        }
    }

    override fun load(dis: DataInput) {
        location = Location(Bukkit.getWorld(dis.readUTF()), dis.readDouble(), dis.readDouble(), dis.readDouble())
    }

    override fun toString(): String {
        return location?.let { loc ->
            "World: ${loc.world?.name ?: "null"} X: ${loc.x} Y: ${loc.y} Z: ${loc.z}"
        } ?: "null"
    }

    override fun getId(): Byte {
        return 100
    }

    override fun getTagName(): String {
        return "TAG_Location"
    }

    override fun copy(): Tag {
        return LocationTag(name, location!!.clone())
    }
}